### DISCUSSION TITLE
Color/Distance Sensor only recognizes WHITE

### DISCUSSION THREAD

**Discussion by Dr-Paul57** [pybricks/support#1450] [Q&A]
I have an issue with my new colour/distance sensor.  It will only recognize WHITE.  Any other colour that I try and the sensor is immediately triggered, regardless of what colour it is pointing at.   I am using it under a train to use the sensor to decide when the train should stop/start etc.  If I use WHITE tiles to trigger it, the code works fine.  Any other colour and the sensor immediately triggers. . .

**Comment by laurensvalk** [comment]
Can you share your full program?

**Comment by Dr-Paul57** [comment]
Here it is.
Not complete but it runs, as long as I use white.  anything else and the motor runs for less than a second.

**Reply by dlech** [reply]
```python
# this code is for a Return Loop track with a long tail and a 'station' in the Loop
# The Train starts from the end of the Tail looking for the Terminus Station
# Stops here for "station_stop_time" then proceeds (slow acceleration on first trip only)
# into the Loop looking for the next sensor where it stops at the Loop Station
# it then proceeds back to the Terminus

#Slow-down time must be less than speed-up time or Slow-Down Markers will be seen incorrectly (All WHITE)
#Would work much better if the detector would 'see' colour!

from pybricks.hubs import CityHub
from pybricks.pupdevices import ColorDistanceSensor, DCMotor # ColorSensor instead of ColorDistanceSensor if using Mindstorms sensor
from pybricks.parameters import Color, Port, Direction
from pybricks.tools import wait

# First check battery:
hub = CityHub()
v = hub.battery.voltage()
i = hub.battery.current()
volts = (v/1000) #convert from milli-volts to volts
# Prints Voltage in 'real' volts to 2 decimal places
print ("voltage =",f"{volts:.2f}", " Volts")
print ('Current =', i, ' milliamps')
# This Number needs to be set to a "Real Value" when I know what it should be!
# Also set Slow_Speed based on Voltage
if volts < 7.8:  # 7.8 is in the BallPark. . . 
    print('Votage Low!!')
    slow_speed = 30  # Starting Speed
else:    
    print ("There's enough Power")
    slow_speed = 30  # Starting Speed

 # Reverse initial motor direction; motor is mounted backwards
motor = DCMotor(Port.A, Direction.COUNTERCLOCKWISE)
 # Set up the Colour Sensor (which seems to only see WHITE)
sensor = ColorDistanceSensor(Port.B)

# This is the Colour Check Function
check_color_interval_ms = 20  # milliseconds between colour check
def check_for_color(color):
    while sensor.color() != color:
        wait(check_color_interval_ms)

# This is the speed_up Function
def speed_up():
    for velocity in range(slow_speed, top_speed, 2):
        forward_speed = velocity
        motor.dc(forward_speed)
        print ("Train Speed = ",forward_speed)
        wait(170)  # Acceleration increment

# This is the slow-down Function
def slow_down():
    for velocity in range(top_speed, slow_speed, -2):
        forward_speed = velocity
        motor.dc(forward_speed)
        print ("Train Speed = ",forward_speed)
        wait(90)  # Deceleration increment
# Acceleration increment MUST be larger than the Deceleration increment!!

# Set up the Initial Parameters
station_stop_time_ms = 8000 # Time before Starting from the Terminus
loop_stop_time_ms = 5000 # Loop Station Stop Time
top_speed = 80  # Top Speed! (Duh!)

# Start the Train Slowly - Looking for Terminus Marker
motor.dc(slow_speed) 
print("looking for Terminus Marker")
check_for_color(Color.WHITE)
# Once the Marker is detected:
print("Terminus Marker detected, \n\t Waiting for 5 Seconds")
motor.stop()
wait(station_stop_time_ms)

#Start the train going
speed_up()

# Now we look for the Slow_Down Marker
print("looking for 'SLOW'")
check_for_color(Color.WHITE)
slow_down()

# Now look for the Loop Station Marker
print("Loop Station Marker detected, \n\t Waiting for 5 Seconds")
check_for_color(Color.WHITE)
motor.stop()
wait(station_stop_time_ms)

#Now start back to Terminus
print("Back to the Terminus")
speed_up()

# Now we look for the Slow_Down Marker
print("looking for 'SLOW'")
check_for_color(Color.WHITE)
slow_down()

print("looking for Terminus Marker")
check_for_color(Color.WHITE)

print("Terminus Reached")
motor.stop()
wait(station_stop_time_ms)
print("\nWorks so far!!")

# Now we have to go the other way without being tripped on the Distand Sensors. . . 

 # Reverse initial motor direction
motor = DCMotor(Port.A, Direction.CLOCKWISE)
#Start rolling
speed_up()
check_for_color(Color.WHITE)
wait(1000) # it's detecting the "forward: slow marker - ignore it!"
# Look for Loop Distant Marker
check_for_color(Color.WHITE)
slow_down
#Check for Loop Station Marker
check_for_color(Color.WHITE)
motor.dc(50) #keep rolling until statin marker
wait(500)
motor.dc(40)
wait(500)
#Look for Loop Station Marker
check_for_color(Color.WHITE)
motor.stop()
wait(loop_stop_time_ms)
#Return to Terminal
speed_up()
#Look for Distance Marker
check_for_color(Color.WHITE)
motor.dc(40)
wait(1800)
motor.stop()
```

**Comment by BertLindeman** [comment]
Hi,

Could you place markdown indicators before and after the python source?
Then we do not have to guess the indentation and such.

Before the source: **\```python**
and after the source just a closing **\```**

It will make the source also better readable.

Thanks,

Bert

**Reply by dlech** [reply]
Unfortunately, that doesn't work when a comment was sent using email - even if you edit it on the web after the fact. I moved it to a new comment to make it easier to read.

**Comment by Dr-Paul57** [comment]
Sorry but I don’t understand what you want me to do!The code indentations and all, read as written on my message containing the code. . . . PJHSent from my iPadOn Feb 17, 2024, at 1:26 PM, Bert ***@***.***> wrote:﻿
Hi,
Could you place markdown indicators before and after the python source?
Then we do not have to guess the indentation and such.
Before the source: ```python
and after the source just a closing ```
It will make the source also better readable.
Thanks,
Bert

**Comment by BertLindeman** [comment]
Sure, no problem. 

Let me try to be more clear:

To make your Python code easier to read and understand, you can format it using GitHub's markdown syntax. 
This will preserve the indentation and syntax highlighting of your code.

Here's how you can do it, by editting your first post on this discussion:

1. On the line before your program, type three backticks followed by the word "python". It should look like this: **\```python** 

2. After your code, type three backticks on a new line to close the code block. It should look like this: **\```**

Here's an example:

\```python
def hello_world():
    print("Hello, world!")
\```

When formatted correctly, it will look like this:

```python
def hello_world():
    print("Hello, world!")
```

Hope this explains it a bit.

Bert

[edit]
You can edit your post by clicking op the top right three dots of your post and choose "edit"

**Comment by Dr-Paul57** [comment]
Ok.   Done on the GitHub post.Dr_PaulSent from my iPadOn Feb 17, 2024, at 4:34 PM, Bert ***@***.***> wrote:﻿
Sure, no problem.
Let me try to be more clear:
To make your Python code easier to read and understand, you can format it using GitHub's markdown syntax.
This will preserve the indentation and syntax highlighting of your code.
Here's how you can do it, by editting your first post on this discussion:


On the line before your program, type three backticks followed by the word "python". It should look like this: ```python


After your code, type three backticks on a new line to close the code block. It should look like this: ```


Here's an example:
```python
def hello_world():
print("Hello, world!")
```
When formatted correctly, it will look like this:
def hello_world():
    print("Hello, world!")
Hope this explains it a bit.
Bert

**Comment by laurensvalk** [comment]
In your code, you have this function:

```python
def check_for_color(color):
    while sensor.color() != color:
        wait(check_color_interval_ms)
```

But throughout the rest of your code, you only have:

```python
check_for_color(Color.WHITE)
```

So it will only look for the color white.

**Comment by Dr-Paul57** [comment]
I have WHITE throughout because that is the ONLY colour that works.   If, for example, I put a red tile on the track and run the routine the train moves for less than a second, stops, waits, runs for less than a second etc.   That is my major issue, I can only use WHITE in the code and WHITE tiles on the track as markers.   No other colour works.Is it the code that’s at fault or the sensor is my main question.   ThanksDr_Paul(Thanks for being patient with me, you’re dealing with a 70 year old who cut his coding skill teeth on a mainframe using Fortran 4!  Program lines typed onto punch cards and only 2 runs a day (none on Sunday!).  So, I am having to learn Python as I go along!)Sent from my iPadOn Feb 18, 2024, at 2:38 AM, laurensvalk ***@***.***> wrote:﻿
In your code, you have this function:
def check_for_color(color):
    while sensor.color() != color:
        wait(check_color_interval_ms)

But throughout the rest of your code, you only have:
check_for_color(Color.WHITE)

So it will only look for the color white.

**Comment by BertLindeman** [comment]
Cobol starter here 😉 

To test the color sensor, you might put the train on it's size and run an example program from the doc, like this:

```python
from pybricks.pupdevices import ColorSensor
from pybricks.parameters import Port, Color
from pybricks.tools import wait

# Initialize the sensor.
sensor = ColorSensor(Port.A)


# This is a function that waits for a desired color.
def wait_for_color(desired_color):
    # While the color is not the desired color, we keep waiting.
    while sensor.color() != desired_color:
        wait(20)


# Now we use the function we just created above.
while True:

    # Here you can make your train/vehicle go forward.

    print("Waiting for red ...")
    wait_for_color(Color.RED)

    # Here you can make your train/vehicle go backward.

    print("Waiting for blue ...")
    wait_for_color(Color.BLUE)
```

Hold some color LEGO blocks near the color sensor and see if it works.

Bert

**Reply by BertLindeman** [reply]
Ah, you use a ColorDistanceSensor on a different port.
I am sure you can adapt my example here or look in the documentation for the similar example for this sensor.

**Comment by laurensvalk** [comment]
We have a tutorial for detecting colors here: https://youtu.be/kwnSbnlhOkA. This is for coding with blocks, but the principles are the same and can also be done in Python. The main thing is that you might need to calibrate your colors. 

Maybe the sensor is very close to the objects you're detecting, so it mostly sees white. 

To apply this in Python, you can see the example syntax for all the equivalent commands in the documentation on the right hand side of the app.

**Comment by Dr-Paul57** [comment]
﻿Hi:I based my code on this one. . .poweredup-color-sensor/pybricks-code-for-60337-express-passenger-train.py at main · eggybricks/poweredup-color-sensorgithub.comI simply did all they did.   The sensor is in the same place, etc.When I ran it using my train the locomotive only moved a tiny bit, stopped, moved a tiny bit, stopped etc.   Then I started trying different colours and found that only WHITE worked.I’ll try your suggestions, move the sensor up one stud and also switch the ‘a’ and ‘b’ channels for the motor and sensor.   I’ll report in once I’ve tried the suggestions.Thanks for all the help I’m getting!Sent from my iPadOn Feb 18, 2024, at 8:11 AM, laurensvalk ***@***.***> wrote:﻿
We have a tutorial for detecting colors here: https://youtu.be/kwnSbnlhOkA. This is for coding with blocks, but the principles are the same and can also be done in Python. The main thing is that you might need to calibrate your colors.
Maybe the sensor is very close to the objects you're detecting, so it mostly sees white.
To apply this in Python, you can see the example syntax for all the equivalent commands in the documentation on the right hand side of the app.

**Reply by Dr-Paul57** [reply]
I had previously watched the tutorial and based some of my work on it.   What I didn’t notice on first watching is that the track in the video is on a white base!   I currently have my track on a cream coloured carpet and, based on the tests I have just run, the sensor sees every colour other than white in the carpet!  If I put white paper under a section of track then RED was detected.   However, that is not practical.   Is there anyway to ‘de-sensitize’ the sensor??   Swapping the A and B channels had no effect, moving the sensor ‘up’ 1 stud was still the same, I couldn’t move it up ‘2’ studs because of the base of the locomotive, the next setting I could use was too far away from the colour tile to be seen.     That’s what I have so far.

**Reply by laurensvalk** [reply]
Please do have a look at [calibrating the colors](https://youtu.be/kwnSbnlhOkA). This technique wasn't available when Eggybricks created their video, but they said they would use this technique next time :)

**Comment by Dr-Paul57** [comment]
Thanks.Is there a way to calibrate the colours with simple Python code or do I have to learn all that ‘block’ stuff?   (Really do NOT want to have to do that!)Dr_PaulSent from my iPadOn Feb 18, 2024, at 1:03 PM, laurensvalk ***@***.***> wrote:﻿
Please do have a look at calibrating the colors. This technique wasn't available when Eggybricks created their video, but they said they would use this technique next time :)

**Reply by laurensvalk** [reply]
As mentioned above, you can do all this with Python too.

You can find examples for all this right inside the app; just go to the documentation for the sensor you are using. You'll see this section:

![image](https://github.com/pybricks/support/assets/12326241/50cdf42d-b1aa-4b01-82d5-422d66e5944c)

**Comment by Dr-Paul57** [comment]
﻿Excellent!Thank you. This will be my ‘project-of-the-day’.By the way;  I also became a ‘Sponsor’ of PyBricks on GitHub - I hope this small monthly contribution helps you continue your excellent work.(Somewhere I read that if I did this on GitHub I should inform you. . . . ?)I’ll report in on the results as soon as I am satisfied that this new method is working!Dr_PaulSent from my iPadOn Feb 19, 2024, at 2:54 AM, laurensvalk ***@***.***> wrote:﻿
As mentioned above, you can do all this with Python too.
You can find examples for all this right inside the app; just go to the documentation for the sensor you are using. You'll see this section:
image.png (view on web)

**Reply by laurensvalk** [reply]
Thanks for becoming a sponsor, much appreciated!

**Reply by Dr-Paul57** [reply]
Hi:  Does this 'sponsorship' get me access to "Block Coding" or do I have to do something on Patreon?
Thanks

**Reply by laurensvalk** [reply]
We've given past sponsors free access since there isn't automatic integration with GitHub.

But since you're a new sponsor, using Patreon or a getting a one-off lifetime license is probably easier. Whichever is most convenient for you. Thank you!

**Comment by Dr-Paul57** [comment]
Download full-resolution images
Available until Mar 20, 2024
 <https://www.icloud.com/attachment/?u=https%3A%2F%2Fcvws.icloud-content.com%2FB%2FAVhww-DEtZe-c2r0dgFqJbNGCoGQAXcPWCLSsVeov_Q9OCwhFwSkdYKc%2F%24%7Bf%7D%3Fo%3DAqrFtxmEBPICLgm5-o5hr1bjFWTLeV4Ea_RS2VYJB4gP%26v%3D1%26x%3D3%26a%3DCAogH9wfbiq-vG-4CNnykV5nyhc1z2Q6FlMz08fQwYN_0WYSeBDn792Q3DEY5__Y5OUxIgEAKgkC6AMA_2h-Wy5SBEYKgZBaBKR1gpxqJvREPc8-Ee0aV-SRtFwhsXwhqVGC4u8RTzX7-mQkPIxLCgkLnMO9ciZpatt9lp43t2-fPybKJbCb_8OUQzJw6S_YEtejdzHO4bumNYLYFA%26e%3D1710950334%26fl%3D%26r%3D8EBB2A0D-9AB5-44B5-A6AA-DE55AD98AE24-1%26k%3D%24%7Buk%7D%26ckc%3Dcom.apple.largeattachment%26ckz%3D721666F9-6711-4E24-B991-0265528EC0C2%26p%3D110%26s%3DKHtou7jLfXe9z4RH1sK-6APUok8&uk=GgK0W7CgsefvWXMYvrY-Ew&f=Images.zip&sz=6763885>Hi:
Well, no luck I’m afraid.

Here are pictures of the test track, white card underneath to reduce the ’noise’.

Then 2 showing how the sensor is mounted and finally a video or the test run.

Here is the code I used:

‘’’PYTHON
from pybricks.hubs import CityHub
from pybricks.pupdevices import DCMotor, Light
from pybricks.parameters import Button, Color, Direction, Port, Side, Stop
from pybricks.robotics import DriveBase
from pybricks.tools import wait, StopWatch

hub = CityHub()

from pybricks.pupdevices import ColorDistanceSensor
from pybricks.parameters import Port
from pybricks.tools import wait

# Initialize the sensor.
sensor = ColorDistanceSensor(Port.B)
motor = DCMotor(Port.A, Direction.COUNTERCLOCKWISE)

motor.dc(20)
while True:
    # Read the color.
    color = sensor.color()

    # Print the measured color.
    print(color)

    # Move the sensor around and see how
    # well you can detect colors.

    # Wait so we can read the value.
    wait(100)
‘’'

And finally, the results as the train moves slowly down the track (see video):

Color.YELLOW
Color.GREEN
Color.BLUE
Color.WHITE
Color.BLUE
Color.BLUE
Color.BLUE
Color.BLUE
Color.GREEN
Color.BLUE
Color.BLUE
Color.BLUE
Color.NONE
Color.GREEN
Color.BLUE
Color.BLUE
Color.BLUE
Color.BLUE
Color.GREEN
Color.WHITE
Color.BLUE
Color.BLUE
Color.GREEN
Color.YELLOW
Color.GREEN
Color.WHITE
Color.BLUE
Color.BLUE
Color.GREEN
Color.GREEN
Color.WHITE
Color.BLUE
Color.BLUE
Color.BLUE
Color.GREEN
Color.GREEN
Color.BLUE
Color.BLUE
Color.BLUE
Color.GREEN
Color.BLUE
Color.WHITE
Color.BLUE
Color.BLUE
Color.GREEN
Color.BLUE
Color.BLUE
Color.BLUE
Color.YELLOW
Color.WHITE
Color.WHITE
Color.WHITE
Color.WHITE
Color.WHITE
Color.WHITE
Color.WHITE
Color.WHITE
Color.WHITE
Color.BLUE
Color.GREEN
Color.BLUE
Color.BLUE
Color.BLUE
Color.BLUE
Color.GREEN
Color.WHITE
Color.BLUE
Color.BLUE
Color.BLUE
Color.BLUE
Color.GREEN
Color.WHITE
Color.BLUE
Color.BLUE
Color.GREEN
Color.GREEN
Color.BLUE
Color.BLUE
Color.GREEN
Color.GREEN
Color.WHITE
Color.BLUE
Color.BLUE
Color.RED
Color.RED
Color.RED
Color.RED
Color.RED
Color.RED
Color.RED
Color.RED
Color.RED
Color.RED
Color.RED
Color.YELLOW
Color.BLUE
Color.BLUE
Color.BLUE
Color.GREEN
Color.GREEN
Color.WHITE
Color.BLUE
Color.BLUE
Color.YELLOW
Color.BLUE
Color.BLUE
Color.BLUE
Color.GREEN
Color.GREEN
Color.BLUE
Color.BLUE
Color.BLUE
Color.NONE
Color.GREEN
Color.WHITE
Color.BLUE
Color.YELLOW
Color.YELLOW
Color.YELLOW
Color.YELLOW
Color.YELLOW
Color.YELLOW
Color.YELLOW
Color.YELLOW
Color.YELLOW
Color.YELLOW
Color.YELLOW
Color.YELLOW
Color.YELLOW
Color.BLUE
Color.BLUE
Color.BLUE
Color.NONE
Color.GREEN
Color.WHITE
Color.BLUE
Color.BLUE
Color.BLUE
Color.YELLOW
Color.WHITE
Color.BLUE
Color.BLUE
Color.GREEN
Color.BLUE
Color.BLUE
Color.BLUE
Color.BLUE
Color.BLUE
Color.YELLOW
Color.BLUE
Color.BLUE
Color.NONE
Color.GREEN
Color.GREEN
Color.GREEN
Color.GREEN
Color.GREEN
Color.GREEN
Color.GREEN
Color.GREEN
Color.GREEN
Color.GREEN
Color.GREEN
Color.YELLOW
Color.GREEN
Color.WHITE
Color.BLUE
Color.BLUE
Color.BLUE
Color.GREEN
Color.WHITE
Color.BLUE
Color.BLUE
Color.NONE
Color.GREEN
Color.BLUE
Color.BLUE
Color.BLUE
Color.BLUE
Color.GREEN
Color.WHITE
Color.BLUE
Color.BLUE
Color.BLUE
Color.GREEN
Color.GREEN
Color.WHITE
Color.BLUE
Color.BLUE
Color.BLUE
Color.BLUE
Color.BLUE
Color.BLUE
Color.BLUE
Color.BLUE
Color.BLUE
Color.BLUE
Color.BLUE
Color.BLUE
Color.BLUE
Color.BLUE
Color.GREEN
Color.GREEN
Color.WHITE
Color.BLUE
Color.BLUE
Color.GREEN
Color.BLUE
Color.BLUE
Color.BLUE
Color.BLUE
Color.BLUE
Color.GREEN
Color.GREEN
Color.WHITE
Color.BLUE
Color.BLUE
Color.GREEN
Color.BLUE
Color.BLUE
Color.BLUE
Color.BLUE
Color.BLUE
Color.NONE
Color.GREEN
Color.BLUE
Color.WHITE
Color.BLUE
Color.BLUE
Color.BLUE
The program was stopped (SystemExit).

The coloured blocks on the track are observed as blocks of ~8-12 repeats.
The order is:  WHITE, RED, YELLOW, GREEN, BLUE

There is so much noise that the train is unable to determine exactly where it is!
I suppose I have to write code where it MUST detect the Colour that I am looking for > 8 times before it accepts it as correct. . 
But of course, that’s going to vary with Train speed. . . .
Any help would be most welcome.

P.S.  I did use your colour detection routing to get the hsv numbers, and then used them in the routing setting “MY” colours; 
Still completely useless.  Do you think that the sensor is defective???

That’s it for now.

Dr_Paul

> On Feb 19, 2024, at 8:01 AM, laurensvalk ***@***.*** ***@***.***>> wrote:
> 
> 
> Thanks for becoming a sponsor, much appreciated!
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/orgs/pybricks/discussions/1450#discussioncomment-8517187>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/BGBHWQX7LT46P4JNSG4OAILYUNEKDAVCNFSM6AAAAABDM6UWPOVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DKMJXGE4DO>.
> You are receiving this because you authored the thread.
> 
>  <http://schema.org/> <https://github.com/orgs/pybricks/discussions/1450#discussioncomment-8517187> <https://github.com/orgs/pybricks/discussions/1450#discussioncomment-8517187> <https://github.com/>Click to Download
 <https://www.icloud.com/attachment/?u=https%3A%2F%2Fcvws.icloud-content.com%2FB%2FAWY-hF6fBRimkk4QlJVpoKhsbAByARcxmbyNUoPZ0QjgpROr4Z7sERAM%2F%24%7Bf%7D%3Fo%3DAjmjxs9bIWsZS2cMv_BGWqLrkdkiY5W39ZGolk5aF6GF%26v%3D1%26x%3D3%26a%3DCAog-RsARpOBuRQBHtew1F8rke0FLY9w6jgOP6d7vLdva00SeBDY7d2Q3DEY2P3Y5OUxIgEAKgkC6AMA_xSGaRZSBGxsAHJaBOwREAxqJlZan89NWJi9tMuLeI3kfbOu_WODPhCEVWjEhYFmk9nL2gfsGeT8ciaxR95o1KWXIR5_D8HtHYQOOo4CTsTMIzPBRpBIogTrwdzutvuoQw%26e%3D1710950334%26fl%3D%26r%3DCC8C92F8-7387-4814-A2EA-D90EDE7868EC-1%26k%3D%24%7Buk%7D%26ckc%3Dcom.apple.largeattachment%26ckz%3D721666F9-6711-4E24-B991-0265528EC0C2%26p%3D110%26s%3DKgp4dJCmZy_HF-2OJ_tbPeyKLIA&uk=6l7FE9Kmsnn2gE2av0EWfg&f=IMG_6410.MOV&sz=36368087>IMG_6410.MOV
36.4 MB

**Reply by laurensvalk** [reply]
What if you run these examples from the documentation?

- [Measuring colors](https://docs.pybricks.com/en/latest/pupdevices/colordistancesensor.html#reading-hue-saturation-value)
- [Setting your own colors ](https://docs.pybricks.com/en/latest/pupdevices/colordistancesensor.html#changing-the-detectable-colors) (using what you've just found)

**Reply by laurensvalk** [reply]
If you share what kind of colors you've measured using the first step, that would be helpful.

**Comment by Dr-Paul57** [comment]
Hi:

I tried both of your code example - no good.

The colours that I measured were the ones actually used on the track.  WHITE, RED, YELLOW, GREEN, BLUE as shown on the picture and the video. . . 
I measured over these tiles, in this orientation, put the numbers into the “Use your Own Colours” example and got the same type of response.
Just simply garbled colours. . . 

Colour	Hue	Saturation	Value
White	52	3	91
Red	359	96	71
Orange	3	97	88
Yellow	52	96	85
Green	130	91	39
Blue	222	96	62
Sleeper (DBG )	190	36	20
Carpet	42	48	34
(DBG = Dark Bluish Gray - the colour of the track sleepers)

I did notice that the V number for what I wanted (not background) was markedly higher than the background, 
But this attempt still resulted on garbage data for colour as the train ran along. . .

Dr_Paul

**Reply by laurensvalk** [reply]
Did you put the two examples together, configuring it to detect your colors? Can you share the code where you did so?

**Comment by Dr-Paul57** [comment]
I think so.I’ll run a repeat tomorrow and send the results.Dr_PaulSent from my iPadOn Feb 19, 2024, at 1:32 PM, laurensvalk ***@***.***> wrote:﻿
Did you put the two examples together, configuring it to detect your colors?

**Reply by laurensvalk** [reply]
The colors look OK, so the combined program may be more relevant to review :)

**Comment by Dr-Paul57** [comment]
﻿So I set it the train up to run over a section of BLUE plates and RED plates with standard track, laid over white card in between.Took about 100 datapoints for each colour and averaged the numbers for h, s & v for both.I then plugged these numbers into a second piece of code and told it to look for ‘my’ BLUE and RED.The train then proceeded to tell me that all the track was blue (it’s dark bluish gray over white card) but it did find the RED.I am working on GREEN, YELLOW & WHITE.Thought I’d better report in, incase you thought I’d given up!(My ‘jobs’ list just got too long to ignore!)ThanksDr_PaulSent from my iPadOn Feb 19, 2024, at 3:19 PM, laurensvalk ***@***.***> wrote:﻿
The colors look OK, so the combined program may be more relevant to review :)

**Reply by laurensvalk** [reply]
> averaged the numbers for h, s & v for both

You'd need to add the color of the track and background below the tracks too. Have a look [at the tutorial](https://www.youtube.com/watch?v=kwnSbnlhOkA) for the rationale behind this.

**Comment by Dr-Paul57** [comment]
﻿Finally it works (in the trial anyway).   I determined the carpet and track colours as well as red, blue, yellow, green and white.Plugged those values into the code and told it to only find those colours.  On the test run, it found the ‘coloured’ tiles as it ran along the track, reporting the colour it was seeing every 1/10th of a second.Now to put these findings into my main program and run around the track backwards and forwards.Thanks SO much for all the help.I’ll be sure to send you the code when I finally complete it so others can benefit from the ‘fruits-of-my-labours’!!Dr_PaulSent from my iPadOn Feb 23, 2024, at 1:00 PM, laurensvalk ***@***.***> wrote:﻿

averaged the numbers for h, s & v for both

You'd need to add the color of the track and background below the tracks too. Have a look at the tutorial for the rationale behind this.

**Reply by laurensvalk** [reply]
I'm glad to hear it! Have fun adding this to your own project.
